src/db/car/forest.rs

```diff
@@ -126,57 +176,76 @@ where
 {
     #[tracing::instrument(level = "trace", skip(self))]
     fn get(&self, k: &Cid) -> anyhow::Result<Option<Vec<u8>>> {
-        let ForestCarInner {
-            reader,
-            frame_cache,
-            write_cache,
-            index,
-            ..
-        } = &mut *self.inner.lock();
         // Return immediately if the value is cached.
-        if let Some(value) = write_cache.get(k) {
+        if let Some(value) = self.write_cache.lock().get(k) {
             return Ok(Some(value.clone()));
         }
 
-        for position in index.lookup(*k)?.into_iter() {
-            let block_map = frame_cache.try_get_or_insert(position, || {
-                // Seek to the start of the zstd frame
-                reader.seek(SeekFrom::Start(position))?;
-                // Decode entire frame into memory
-                let mut zstd_frame = decode_zstd_single_frame(reader)?;
-                // Parse all key-value pairs and insert them into a map
-                let mut block_map = HashMap::new();
-                while let Some(block_frame) = UviBytes::default().decode_eof(&mut zstd_frame)? {
-                    if let Some(Block { cid, data }) = Block::from_bytes(block_frame) {
-                        block_map.insert(cid, data);
-                    } else {
-                        return Err(invalid_data("corrupted key-value block"));
+        let mut indexed_guard = self.indexed.lock();
+        let indexed = hoist_error(&mut *indexed_guard)?;
+
+        for position in indexed.lookup(*k)?.into_iter() {
+            let reader = indexed.get_mut();
+            let cache_query = self.frame_cache.lock().get(position, self.cache_key, *k);
+            match cache_query {
+                // Frame cache hit, found value.
+                Some(Some(val)) => return Ok(Some(val)),
+                // Frame cache hit, no value. This only happens when hashes collide
+                Some(None) => {}
```

Comment:
Do we expect hash collisions? If not, should we panic here?

Reply:
Hash collisions may occur, and it costs us little to handle them correctly. If two CIDs have the same hash, the index returns two possible positions. We check the second position if we don't find the correct CID at the first position.